types/scenegraph.d.ts

```diff
@@ -10,6 +10,122 @@ declare interface ScaleFactor {
+ *
+ * *Each* scenenode has its own metadata storage, accessed via `SceneNode.sharedPluginData`. To store general metadata that is not specific to one scenenode, use `sharedPluginData` on the document's scenegraph root.
+ * @example ```js
+ * // This example shows how to save & retrieve rich JSON data in shared metadata storage.
+ // See below for simpler examples of using individual APIs.
+ const PLUGIN_ID = "<your manifest's plugin ID here>";
+ let richObject = {
+    list: [2, 4, 6],
+    name: "Hello world"
+ };
+ node.sharedPluginData.setItem(PLUGIN_ID, "richData", JSON.stringify(richObject));
+
+ // Later on...
+ // (This could be in a different plugin, if it passes the original plugin's ID here)
+ let jsonString = node.sharedPluginData.getItem(PLUGIN_ID, "richData");
+ if (jsonString) {  // may be undefined
+    let richObjectCopy = JSON.parse(jsonString);
+    console.log(richObjectCopy.list.length);  // 3
+ }
+ */
+declare interface PerPluginStorage {
+    /**
+     * Returns a map where key is plugin ID and value is a nested map containing all the shared metadata for that plugin ID (i.e. the result of calling `getForPluginId()` with that ID).
+     */
+    getAll(): { [key: string]: {[key:string]: string} };
+
+    /**
+     * Returns a map of key-value string pairs containing all shared metadata stored on this node by the given plugin. May be an empty object (zero keys), but is never null.
+     *
+     * This map is a clone of the stored metadata, so modifying it has no effect.
+     * @param pluginId
+     *
+     * @example
+     *const MY_PLUGIN_ID = "<your manifest's plugin ID here>";
+let mySharedMetadata = node.sharedPluginData.getForPluginId(MY_PLUGIN_ID);
+console.log("My shared 'foo' & 'bar' values:",
+    mySharedMetadata.foo, mySharedMetadata.bar);
+
+
+console.log("Plugin B's shared 'foo' value:",
+    node.sharedPluginData.getForPluginId("B").foo);
+     */
+    getForPluginId(pluginId: string): {[key:string]: string};
+
+    /**
+     * Returns a list of all keys stored on this node by the given plugin. May be empty (length zero), but is never null.
+     * @param pluginId
+     *
+     * @example
+     * console.log("All properties stored by plugin A on this node:",
+     node.sharedPluginData.keys("A"));
+     */
+    keys(pluginId: string): string[];
+
+    /**
+     * Returns the value stored under the given key on this node by the given plugin, or `undefined` if the plugin hasn't stored anything under the given key.
+     *
+     * Because metadata is stored separately per plugin, two plugins can store two different values under the same key.
+     * @param pluginId
+     * @param key
+     *
+     * @example
+     * // These are two different values, stored independently per plugin
+ console.log("Plugin A's 'foo' value:", node.sharedPluginData.getItem("A", "foo"));
+ console.log("Plugin B's 'foo' value:", node.sharedPluginData.getItem("B", "foo"));
```

Comment:
Formatting.

```suggestion
     * @example
     * ```
     * // These are two different values, stored independently per plugin
     * console.log("Plugin A's 'foo' value:", node.sharedPluginData.getItem("A", "foo"));
     * console.log("Plugin B's 'foo' value:", node.sharedPluginData.getItem("B", "foo"));
     * ```
```